### ISSUE TITLE
findRecord does not respect reload option when id string does not match result

### ISSUE THREAD

**Issue by achidlow** [emberjs/data#4414]
I'm using token authentication, and have a REST endpoint /users/me that returns the record for the current user based on the authorization token.

When logging out one user and logging in another, even though `findRecord` is called with `{reload: true}`, it doesn't hit up the API.

I've tried various workarounds, the best I could come up with is to find the record with key 'me' in the store and unload it manually.

I've tried stepping through the source code to figure out why this is happening, but I'm very new to front-end development. It looks like the record gets stuck in the `loading` state, so you can't `peekRecord()` it directly either.

``` javascript
import Ember from 'ember';
import DS from 'ember-data';
import ESASession from "ember-simple-auth/services/session";

export default ESASession.extend({
    store: Ember.inject.service(),

    user: Ember.computed('isAuthenticated', function()
    {
        if (this.get('isAuthenticated')) {
            let store = this.get('store');
            // this shouldn't be necessary! ember-data for some reason doesn't like the 'me'
            // key, and the reload option isn't being respected
            const me = store.peekAll('user').findBy('id', 'me');
            if (me) {
                store.unloadRecord(me);
            }
            const promise = store.findRecord('user', 'me', {reload: true});
            return DS.PromiseObject.create({ promise: promise });
        }
    })
});
```

**Comment by bmac** [comment]
@achidlow This sounds like a bug. I assume at some point when you do a `store.findRecord('user', 'me', {reload: true});` your api is returning a response to indicate that there is no `user/me` record. Is that correct?

**Comment by achidlow** [comment]
@bmac thanks for your response!

I've just double checked and this occurring even in the case that `/users/me` only ever returns a user. So for example:
1. The page is initially loaded with the user already logged in (I'm using ember-simple-auth so the user token is persisted to localstorage), which makes a call to `/users/me` to get user info and successfully retrieves user.
2. The user log outs, which doesn't make any API calls, 
3. A different user logs in, which should be making a call to `/users/me` because of the `reload` option, but isn't. 

You can see the above in this screenshot, which was from running without my `unloadRecord` workaround:
![reloadnotworking](https://cloud.githubusercontent.com/assets/5511888/15767414/4c47b210-297a-11e6-800d-dec21a961593.png)

**Comment by pangratz** [comment]
I suspect you're running into the issue where you are requesting a record with id `me` but your server returns a record with a different id (the real one).

The problem is that now there are 2 records within the store:

- a record with id returned by the server (in the loaded state)
- a record with id `me` (still in the loading state since no data for a record with that id has been pushed to the store)

Ember-Data expects the id of the primary data of the response to match the id of the requested record. Since there is a mismatch, Ember-Data gets confused.

The most straightforward solution IMHO here would be to use `store.queryRecord('user', { me: true })` (or the like) and overwrite your adapters' `urlForQueryRecord` method to return `/api/users/me` when `query.me` is `true`. By using `queryRecord` you will always ask the server for the freshest data. See [this ember-twiddle](https://ember-twiddle.com/7908f3a1d3c1164ca70ae0290042a51d?fileTreeShown=false&numColumns=3&openFiles=routes.application.js%2Cadapters.application.js%2Ccontrollers.application.js) for an example.

**Comment by wecc** [comment]
@pangratz should this be closed since #4604 has been merged?

**Comment by pangratz** [comment]
Good pount @wecc, closing since #4604 is merged and should provide a better feedback.
